python/ray/rllib/es/policies.py

```diff
@@ -5,191 +5,85 @@
+                                           ".max_episode_steps")
+    timestep_limit = (env_timestep_limit if timestep_limit is None
+                      else min(timestep_limit, env_timestep_limit))
+    rews = []
+    t = 0
+    observation = env.reset()
+    for _ in range(timestep_limit):
+        ac = policy.compute(observation, add_noise=add_noise)[0]
+        observation, rew, done, _ = env.step(ac)
+        rews.append(rew)
+        t += 1
+        if done:
+            break
+    rews = np.array(rews, dtype=np.float32)
+    return rews, t
+
+
+class GenericPolicy(object):
+    def __init__(self, sess, action_space, preprocessor,
+                 observation_filter, action_noise_std):
+        self.sess = sess
+        self.action_space = action_space
+        self.action_noise_std = action_noise_std
         self.preprocessor = preprocessor
 
-        with tf.variable_scope(type(self).__name__) as scope:
-            # Observation normalization.
-            ob_mean = tf.get_variable(
-                'ob_mean', self.preprocessor.shape, tf.float32,
-                tf.constant_initializer(np.nan), trainable=False)
-            ob_std = tf.get_variable(
-                'ob_std', self.preprocessor.shape, tf.float32,
-                tf.constant_initializer(np.nan), trainable=False)
-            in_mean = tf.placeholder(tf.float32, self.preprocessor.shape)
-            in_std = tf.placeholder(tf.float32, self.preprocessor.shape)
-            self._set_ob_mean_std = U.function([in_mean, in_std], [], updates=[
-                tf.assign(ob_mean, in_mean),
-                tf.assign(ob_std, in_std),
-            ])
-
-            inputs = tf.placeholder(
-                tf.float32, [None] + list(self.preprocessor.shape))
-
-            # TODO(ekl): we should do clipping in a standard RLlib preprocessor
-            clipped_inputs = tf.clip_by_value(
-                (inputs - ob_mean) / ob_std, -5.0, 5.0)
-
-            # Policy network.
-            dist_class, dist_dim = ModelCatalog.get_action_dist(
-                self.ac_space, dist_type='deterministic')
-            model = ModelCatalog.get_model(clipped_inputs, dist_dim)
-            dist = dist_class(model.outputs)
-            self._act = U.function([inputs], dist.sample())
-        return scope
-
-    def act(self, ob, random_stream=None):
-        a = self._act(ob)
-        if not isinstance(self.ac_space, gym.spaces.Discrete) and \
-                random_stream is not None and self.ac_noise_std != 0:
-            a += random_stream.randn(*a.shape) * self.ac_noise_std
-        return a
-
-    @property
-    def needs_ob_stat(self):
-        return True
-
-    @property
-    def needs_ref_batch(self):
-        return False
-
-    def set_ob_stat(self, ob_mean, ob_std):
-        self._set_ob_mean_std(ob_mean, ob_std)
+        if observation_filter == "MeanStdFilter":
+            self.observation_filter = MeanStdFilter(
+                self.preprocessor.shape, clip=None)
```

Comment:
Should this be kept synchronized among workers somewhere? I see that before there was an _set_ob_stat function which is sometimes called, but now we just set theta.

Reply:
We should definitely push this into a Catalog

Reply:
@richardliaw Agreed about the catalog.

@ericl As for the actual algorithm, you're right this is a change in behavior from before. I think that before, statistics were only being collected once in a while (one in every hundred times, that's what the `calc_obstat_prob=0.01` value was), returned to the master, aggregated, and set on all the workers. Now we update the observation filter with every step but we don't share the values with the other workers. The upsides are that it's simpler to implement, though it does make things more "stateful". I'm guessing it doesn't hurt performance too much.